### ISSUE TITLE
Implement a state machine in Fulfillment

### ISSUE THREAD

**Issue by jonyw4** [vendure-ecommerce/vendure#426]
We already have a state machine in Order, Payments and Refunds, I believe that we should implement that in the Fulfillment too so that we can make things like: create labels for shipping, create tracking code (this depends on carrier), notify users of the state of the fulfillment by email and many other things...

## Features
- Implement a state machine in Fulfillment. I thought in these states initially:  ` 'Pending' | 'Shipped' | 'Delivered' | 'Cancelled'`
- Button in Admin UI to change the Fulfillment state (like Payment and Refunds)
- A way to define how the Fulfillment should work when transiting between states

**Comment by jonyw4** [comment]
I create a PR about this issue that is WIP, so then we can discuss how this should work

**Comment by michaelbromley** [comment]
Thanks for taking the time to think about this issue. Some feedback from me:

1. Since we will be using the state machine just like Orders, we should consider using a similar method of extending the states & reacting to transitions, so something equivalent to [CustomOrderProcess](https://github.com/vendure-ecommerce/vendure/blob/master/packages/core/src/config/order/custom-order-process.ts)
2. Your proposed states include `Refunded`, which to me seems like the concern of a Payment rather than a Fulfillment. I would propose: `'Pending' | 'Shipped' | 'Delivered' | 'Cancelled'`.
3. This also leads me to think about handling of returns, but I think returns should be separate entities, since it is likely that only _part_ of a fulfillment would be returned commonly, so it makes sense to track that as a separate entity with its own state. So out of the scope of this feature, albeit related.

**Comment by jonyw4** [comment]
Nice @michaelbromley, I agree with everything you said!

**Comment by jonyw4** [comment]
I had progress in this PR, now in Vendure is possible to define a fulfillmentOption that can have a custom Fulfillment process like you describe. I change the states like you propose too.

My next steps will be:
- Implement `ShippingFulfillmentMethod` class
- Add `ShippingFulfillmentMethod` class in vendure config fulfillmentOption
- Add E2E Fulfillment Proccess tests
- Add docs about custom fulfillment proccess

@michaelbromley I am learning Angular yet, so I can't make changes in the Admin UI. I believe that won't be hard to do because will be very similar with order / payment UI. So, can you create the UI for this PR?

**Comment by michaelbromley** [comment]
Hey Jonthan, this is amazing work so far. A pretty complex task you are taking on but you're definitely moving in the right direction!

> Implement ShippingFulfillmentMethod class
Can you elaborate a bit on what this class is for and how it will work?

>  So, can you create the UI for this PR?
Yes, you just worry about the server side parts. For me the UI part should not be too much extra work.

**Comment by jonyw4** [comment]
I don't know the name and methods yet, but the ideia is to have a class similar to `ShippingCalculator` or `PaymentMethod` called `FulfillmentMethod` where user can define how he want to handle the fulfillment. We can add in this class methods like `ship`, `stopShipping`. This class should be defined in Vendure Config, and them in Admin UI the user can config this `FulfillmentMethod`

**Comment by michaelbromley** [comment]
OK, I think I get the idea.

This is quite a large scope for a single PR. I'd suggest if possible to limit the initial work to establishing the underlying state machine mechanism for Fulfillments. Then in a separate PR you build on it with the rest of what you described. Otherwise it'll be harder to track all the changes.

**Comment by dimeros** [comment]
So, can somebody else work on the admin UI in parallel? What is it expected? The states of fullfillment option can be transitioned from the UI such as in "Orders" case?

**Comment by michaelbromley** [comment]
@dimeros I'd suggest that the server-side groundwork be completed first before anyone starts on the Admin UI aspect. I'm still not 100% clear on @jonyw4's plans on the design, so I'm just waiting to see any further commits to the PR #427

**Comment by jonyw4** [comment]
@michaelbromley, [I create another issue just to discuss custom fulfillment methods.](https://github.com/vendure-ecommerce/vendure/issues/447)

Here in this issue and in the #427 PR I will discuss just about add a state machine in Fulfillment Method for maintain the current scope

**Comment by jonyw4** [comment]
@dimeros About the admin UI transition, I think that we can have a button like Refunds and Payment to transit between the Fulfillment state. And then we could implement a class to define what the fulfillment should do when transit to a specific state. I create the #447 to discuss that

**Comment by jonyw4** [comment]
Hey @michaelbromley, we can have also a way 

Ful
Shipped
Delivered

**Comment by jonyw4** [comment]
Sorry, I realize that if we will have CustomFulfillmentProcess, we will not need to implement this issue #447.

**Comment by jonyw4** [comment]
Hey @michaelbromley, today in Order we have `Fulfilled` and `PartiallyFulfilled` but with creation of new Fulfillment states I thought in maybe change the defaults states in Order to cover these new states. 

Using `PaymentMethod` as example and the current default Order states, we can have these new states:

### `Shipped`
To reach this state all Fulfillment must be in `Shipped` state.

### `PartiallyFulfilled` _(or maybe change to `PartiallyDelivered`)_
To reach this state, the Order must have at least one  Fulfillment in `Delivered` state. We can also remove this state.

### `Fulfilled` _(or maybe change to `Delivered`)_
To reach this state all Fulfillment must be in `Delivered` state.

--- 

I think that we should change the names of the old states to be more accurate, but I don't know if this could break something in Vendure...

**Comment by michaelbromley** [comment]
@jonyw4 yes, I agree with all of the above. Also changing the state names to "delivered". Yes, it'll be a breaking change but hopefully TypeScript will help to avoid uncaught breakages.

**Comment by jonyw4** [comment]
Okay I will change that in the PR.

**Comment by jonyw4** [comment]
I little update here, when I started to update the order e2e tests I realize that we will need `PartiallyShipped` order state too.
Today If you forget to fulfill an order item, the order state will continue in `PartiallyFulfilled`. We need to achieve that in this PR too. So the new order states must be that:

### `PartiallyShipped`
To reach this state, the Order must have at least one Fulfillment in `Shipped` state.

### `Shipped`
To reach this state all Fulfillment must be in `Shipped` state **and all order items must have a Fulfillment**

### `PartiallyDelivered` *(old `PartiallyFulfilled`)*
To reach this state, the Order must have at least one  Fulfillment in `Delivered` state.

### `Delivered` *(old `Fulfilled`)*
To reach this state all Fulfillment must be in `Delivered` state **and all order items must have a Fulfillment**

**Comment by michaelbromley** [comment]
### Status update

The PR implementing the server-side of this is now merged into `next`. 

Tasks outstanding:

- [ ] Admin UI parts to interact with Fulfillment state
- [ ] Documentation
